meeting-notes/2022-05-26.md

```diff
@@ -0,0 +1,115 @@
+  I don't think it's removing the old hats properly.
+    - You can tell because scrolling the editor slightly (which will trigger
+      JetBrains to redraw everything, but not enough to actually change the
+      hats) "fixes" things.
+- There's no way to run commands yet.
+
+## What's next for the prototype?
+
+See MILESTONES.md for most of it.
+
+## Notes / learnings:
+
+- VS Code plugins are single threaded (VS Code itself has multiple JS engines.
+  The editor has its own, but all plugins share a single one)
+- Extension interoperation: extensions can return stuff from `activate()` which
+  could be called by other extensions. See `getExtensionApi.ts`
+    - This is pretty handy, because we can have the sidecar extension (better
+      name needed) and Cursorless talked to each other as needed.
+- Every VS Code window is its own instance (with its own command server). It can
+  be difficult to figure out which window is focused/frontmost.
+    - If they ran HTTP servers, each one would need its own port.
+    - Again, probably not a problem for the Cursorless Everywhere user since
+      they won't be using VS Code otherwise.
+        - Although if we will have mixed users, we could consider running a
+          special version of VS Code just for the sidecar.
+- Cursorless commands use async/await and return promises which _do not fulfil
+  until VS Code has updated all of the listeners registered about those
+  changes_ (cool!)
+    - Thus, it might be possible to accumulate all the diffs that were generated
+      during the course of running a single cursorless command, because of the
+      single threaded aspect of plugins + promise contract.
+- Therefore -- whoever triggers the invocation of the cursorless command
+  (via command server?) could also collect all of the editor diffs associated
+  with that command, and send them to JetBrains.
+
+## Medium term
+
+- `addDecorationsToEditors`: this is a nicely abstractable function that we
+  could use as the first piece to break out of the VS Code dependency.
+    - It could basically be run in isolation to "generate hats on this
+      in-memory, abstracted editor state and return them to me as a dictionary".
+    - This isn't very useful for us in the short term however, because we
+      actually _do_
+      need the hat state to be persisted everywhere -- other pieces of
+      Cursorless functionality depend on it -- so we're not going to pursue
+      this.
+    - @pokey mention that it is likely that this gets broken out though, because
+      rango needs it.
+
+## How should commands get to VS Code?
+
+Two worlds:
+
+1. commands go to VS Code (this is how it works) **(world one)**
+    - this is simplest to get started. We simply add `app: jetbrains` to
+      cursorless.talon, and then make the transmission of commands work without
+      VS Code focused (see MILESTONES.md)
+    - it would be nice if we could block the cursorless talon action on the
```

Comment:
fwiw I think this isn't sufficient to guarantee strict serialization of talon actions.  It will guarantee that cursorless actions preceed keystrokes, but not the other way around